staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go

```diff
@@ -507,6 +507,17 @@ func (a *Authenticator) refreshCredsLocked(r *clientauthentication.Response) err
 	expiry := time.Time{}
 	if a.cachedCreds.cert != nil && a.cachedCreds.cert.Leaf != nil {
 		expiry = a.cachedCreds.cert.Leaf.NotAfter
+
+		// this method will be invoked twice for the same request if we consider the
+		// expirationTimestamp to be in the past.  if this happens when cert based
+		// auth is being used and the exec plugin returns a new cert for each invocation,
+		// we will accidentally close the in-flight connection because we think that the
+		// cert rotated (even though we never used the first cert at all).
+		if a.credsExpiredLocked() {
+			klog.InfoS("certificate was marked expired when issued, ignoring expiration until unauthorized",
+				"expirationTimestamp", a.exp.String())
+			a.exp = time.Time{}
```

Comment:
while this may happen to drive the behavior you want, it's pretty hard to reason about and doesn't seem correct to me

* it only gets set for cert-based auth
* it resets a.exp which is set above (which is hard to follow)
* it means credsExpired() will now return false, even though the time returned by the plugin has passed

are you wanting to make sure a credential returned from a plugin with an expirationTimestamp in the past gets used *at least once*, or that returning an expiration timestamp in the past is the same as not return an expiration timestamp at all?

Reply:
Stepping back a bit, I think it may be simpler to frame this as:

For a particular round trip, the exec plugin must only be invoked once and only at the start of the round trip (before the base round tripper is called).

This would allow an exec plugin to return an expiration timestamp in the past, and all that would mean is that the exec plugin would be invoked on every single round trip (i.e. no caching).

This would work for both forms of auth without resetting a.exp and thus credsExpired would behave as before.

Since the exec plugin would only be called once at the start, it would be impossible for it to close the in-flight connection, because that connection does not start until the base round tripper is called.

Reply:
I took a shot at implementing the behavior I described above.  I am not happy with the change per se, but it seems better than what we have today.  Added tests that exercise the edge case.